src/dag.rs

```diff
@@ -1,11 +1,155 @@
 use crate::error::Error;
 use crate::ipld::{decode_ipld, encode_ipld, Ipld};
-use crate::path::IpfsPath;
+use crate::path::{IpfsPath, SlashedPath};
 use crate::repo::RepoTypes;
 use crate::Ipfs;
 use bitswap::Block;
 use cid::{Cid, Codec, Version};
+use ipfs_unixfs::{
+    dagpb::{wrap_node_data, NodeData},
+    dir::{Cache, ShardedLookup},
+    resolve, MaybeResolved,
+};
+use std::convert::TryFrom;
+use std::error::Error as StdError;
+use std::iter::Peekable;
+use thiserror::Error;
 
+#[derive(Debug, Error)]
+pub enum ResolveError {
+    /// Loading of the block on the path failed
+    #[error("block loading failed")]
+    Loading(Cid, #[source] crate::Error),
+
+    /// Document was unsupported; this can be an UnixFs directory structure which has unsupported
+    /// options, or IPLD parsing failed.
+    #[error("unsupported document")]
+    UnsupportedDocument(Cid, #[source] Box<dyn StdError + Send + Sync + 'static>),
+
+    /// Path contained an index which was out of range for the given [`Ipld::List`].
+    #[error("list index out of range 0..{elements}: {index}")]
+    ListIndexOutOfRange {
+        /// The document with the mismatched index
+        document: Cid,
+        /// The path up until the mismatched index
+        path: SlashedPath,
+        /// The index in original path
+        index: usize,
+        /// Total number of elements found
+        elements: usize,
+    },
+
+    /// Path attempted to resolve through for example string or integer.
+    #[error("tried to resolve through object that had no links")]
+    NoLinks(Cid, SlashedPath),
+
+    /// Path attempted to resolve through a property, index or link which did not exist.
+    #[error("no link named {:?} under {0}", .1.iter().last().unwrap())]
+    NotFound(Cid, SlashedPath),
+}
+
+#[derive(Debug, Error)]
+pub enum UnexpectedResolved {
+    #[error("path resolved to unexpected type of document: {:?} or {}", .0, .1.cid())]
+    UnexpectedCodec(cid::Codec, ResolvedNode),
+    #[error("path did not resolve to a block on {}", .0.cid())]
+    NonBlock(ResolvedNode),
+}
+
+/// Used internally before translating to ResolveError at the top level by using the IpfsPath.
+#[derive(Debug)]
+enum RawResolveLocalError {
+    Loading(Cid, crate::Error),
+    UnsupportedDocument(Cid, Box<dyn StdError + Send + Sync + 'static>),
+    ListIndexOutOfRange {
+        document: Cid,
+        segment_index: usize,
+        index: usize,
+        elements: usize,
+    },
+    InvalidIndex {
+        document: Cid,
+        segment_index: usize,
+    },
+    NoLinks {
+        document: Cid,
+        segment_index: usize,
+    },
+    NotFound {
+        document: Cid,
+        segment_index: usize,
+    },
+}
```

Comment:
The use of `{ ... }` makes for quite verbose formatting by rustfmt at the moment. I did this originally with tuples but then got confused about the "count" vs. "index" for some amount of time, and thought that the index prefix might clarify it. Tests should call resolve_local and name the second tuple member `matched` or `matched_count` or similar.